### ISSUE TITLE
The embedder should be handling prompts and alerts

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#19992]
As of now, alerts are handled by browserhtml via a mozbrowser event, or we use tinyfiledialogs in the script thread. mozbrowser is going away.

We should not use tinyfiledialogs in libservo, but in /ports/.

**Comment by fabricedesre** [comment]
How do you see that work? Using an additional method in the embedder and let the embedder send back a message to unblock the script thread execution?

**Comment by paulrouget** [comment]
Yes. See how `allow_navigation` works for example.

**Comment by jdm** [comment]
Relevant code:
* https://github.com/servo/servo/blob/34f388229df4f201302ee51a8969c0556692bf0b/components/constellation/constellation.rs#L1845-L1851 (example of communicating with the embedder that can be used as a model)
* https://github.com/servo/servo/blob/34f388229df4f201302ee51a8969c0556692bf0b/components/script/dom/window.rs#L544-L550 (this already communicates with the constellation to decide whether to show a prompt, then shows the prompt)
* https://github.com/servo/servo/blob/34f388229df4f201302ee51a8969c0556692bf0b/components/script/dom/permissions.rs#L311-L329 (this needs to communicate with the embedder via the constellation)

I'll file separate issues for the uses in components/bluetooth and components/net, since they need to be solved differently. I'm providing this information so some specific contributors can work on this issue; please check with me first before starting to work on it if I have not discussed it with you.

**Comment by gterzian** [comment]
In the light of #20329, where Script would need to ask the embedder to allow unloading of a document: 

1. I starting looking into it and realized script doesn't have a channel to the embedder yet, and also that script doesn't seem to be aware of it's own top level browsing context.
2. I assume constellation shouldn't block waiting for the response from the embedder.
3. In the light of the above, it seems we will have to send a first message from script to constellation, containing a response sender, and then constellation could send a message to the embedder containing the top level browsing context id, and the response sender. The embedder could then directly respond to script via this response sender. 

Does this architecture makes sense? Should we instead add a `EmbedderProxy` directly on script_thread and ask for the top level browsing context id from constellation?

I saw that currently the `Alert` message also isn't yet forwarded to the embedder, see
https://github.com/servo/servo/blob/master/components/constellation/constellation.rs#L1757
So I propose a separate PR to deal with all of this. Sounds good?

**Comment by paulrouget** [comment]
A very similar problem is being addressed here: #20480 and see https://github.com/servo/servo/pull/20480#issuecomment-378306624

**Comment by paulrouget** [comment]
I had a quick chat with @cbrewster and we agree that it's fine if the script thread has access to the EmbedderProxy. We could short-circuit the constellation.
